crates/sui-framework/packages/dwallet-system/sources/dwallet.move

```diff
@@ -0,0 +1,238 @@
+        session_id: ID,
+        dwallet_id: ID,
+        dwallet_cap_id: ID,
+        messages: vector<vector<u8>>,
+        sender: address,
+        sign_data_event: E,
+    }
+
+    // <<<<<<<<<<<<<<<<<<<<<<<< Events <<<<<<<<<<<<<<<<<<<<<<<<
+
+    struct DWalletCap has key, store {
+        id: UID,
+    }
+
+    struct ApprovalsHolder has key {
+        id: UID,
+        message_approvals: vector<MessageApproval>,
+    }
+
+    struct MessageApproval has store {
+        dwallet_cap_id: ID,
+        message: vector<u8>,
+    }
+
+    struct SignMessages<S: store, E: store + copy + drop> has key, store {
+        id: UID,
+        dwallet_id: ID,
+        dwallet_cap_id: ID,
+        messages: vector<vector<u8>>,
+        sign_data: S,
+        sign_data_event: E,
+    }
+
+    struct SharedSignMessages<S: store, E: store + copy + drop> has key {
+        id: UID,
+        sign_messages: SignMessages<S, E>,
+    }
+
+    struct SignSession<S: store> has key {
+        id: UID,
+        dwallet_id: ID,
+        dwallet_cap_id: ID,
+        messages: vector<vector<u8>>,
+        sender: address,
+        sign_data: S,
+    }
+
+    #[allow(unused_field)]
+    struct SignOutput has key {
+        id: UID,
+        session_id: ID,
+        dwallet_id: ID,
+        dwallet_cap_id: ID,
+        signatures: vector<vector<u8>>,
+        sender: address,
+    }
+
+    public(friend) fun create_dwallet_cap(ctx: &mut TxContext): DWalletCap {
+        DWalletCap {
+            id: object::new(ctx),
+        }
+    }
+
+    public fun create_approvals_holder(message_approvals: vector<MessageApproval>, ctx: &mut TxContext) {
+        let holder = ApprovalsHolder {
+            id: object::new(ctx),
+            message_approvals,
+        };
+        transfer::transfer(holder, tx_context::sender(ctx));
+    }
+
+    public fun remove_approvals_holder(holder: ApprovalsHolder): vector<MessageApproval> {
+        let ApprovalsHolder {
+            id,
+            message_approvals,
+        } = holder;
+        object::delete(id);
+        message_approvals
+    }
+
+    public fun approve_messages(dwallet_cap: &DWalletCap, messages: vector<vector<u8>>): vector<MessageApproval> {
+        let dwallet_cap_id = object::id(dwallet_cap);
+        let message_approvals = vector::empty<MessageApproval>();
+        while(vector::length(&messages) > 0) {
+            let message = vector::pop_back(&mut messages);
+            vector::push_back(&mut message_approvals, MessageApproval {
+                dwallet_cap_id,
+                message,
+            });
+        };
+        message_approvals
+    }
+
+    public fun dwallet_cap_id(message_approval: &MessageApproval): ID {
+        message_approval.dwallet_cap_id
+    }
+
+    public fun message(message_approval: &MessageApproval): vector<u8> {
+        message_approval.message
+    }
+
+    public fun remove(message_approval: MessageApproval): (ID, vector<u8>) {
```

Comment:
since this is a module that serves multiple objects, I think it's better to have fully-qualified names like `remove_message_approval`. But let's follow standards if there are.

Reply:
Also note that we have `message` and then `sign_messages_messages`. This is confusing. Let's have fully quallified names for all